server.js

```diff
@@ -5,29 +5,94 @@ const path = require('path');
 const express = require('express');
 const helmet = require('helmet');
 const bodyParser = require('body-parser');
+const compression = require('compression');
 const crypto = require('crypto');
+const morgan = require('morgan');
+const package = require('./package.json');
 
 const config = require('./config.server.js');
 
 const buildDir = 'dist';
 const staticDir = path.join(__dirname, buildDir);
+let indexHTML = '<html></html>';
+
+/**
+ * Check if one file can be found, recursively using paths array
+ * @param {string[]} paths array of paths
+ * @returns {Promise<string|null>} true if found
+ */
+function foundFileFromPath(paths) {
+  return new Promise((resolve) => {
+    if (paths.length < 1) {
+      return resolve(null);
+    }
+    const url = paths.join('/');
+    const localFile = `${staticDir}/${url}`;
+    fs.access(localFile, fs.constants.R_OK, (err) => {
+      if (err) {
+        paths.shift();
+        resolve(foundFileFromPath(paths));
+      } else {
+        resolve(url);
+      }
+    });
+  });
+}
 
-const app = express();
-
-const nonceMiddleware = (req, res, next) => {
-  // Cache static html file to avoid reading it from the filesystem on each request
-  if (!global.html) {
-    console.log('Caching static HTML');
-    global.html = fs.readFileSync(`${staticDir}/index.html`, 'utf8');
+/**
+ * Verify we have the requested file in stock
+ * If we have, but with the wrong path, do a redirect
+ * If not return the modified index.html
+ * @param {Express.Request} req
+ * @param {Express.Response} res
+ * @param {(err?: any) => void} next
+ */
+function redirectMiddleware(req, res, next) {
+  const reqURL = req.url;
+  const urlPath = reqURL.split('/');
+  const n = urlPath.length;
+  const paths = [];
+  for (let i = 0; i < n; i++) {
+    const p = urlPath[i];
+    // Keep only not empty elements, but avoid problematics
+    // paths for security reason.
+    if (p.length > 0 && p != '.' && p != '..') {
+      paths.push(p);
+    }
   }
 
+  foundFileFromPath(paths).then((url) => {
+    if (url) {
+      if (reqURL === `/${url}`) {
+        // Serve the file with 'serve-static'
+        return next();
+      }
+      return res.redirect(`/${url}`);
+    }
+    // Not found, send the modified index.html
+    res.setHeader('Cache-Control', 'public, max-age=0');
+    res.send(res.locals.htmlWithNonces);
+  });
+}
```

Comment:
- We don't need to find files recursively, there are no sub directories in dist folder.
- I think it is safer and clearer to use native path module to get absolute path of the file requested instead of splitting/filtering/joining strings.

For example (with an url rewrite) :
```javascript
/**
 * Check if one file can be found in static dist folder
 * @param {string} urlPath urlPath to match
 * @returns {Promise<string|null>} true if found
 */
function foundFileFromPath(urlPath) {
  return new Promise((resolve) => {
    const file = path.basename(urlPath)
    if(file === "index.html") {
      return resolve(null);
    }
    // Getting normalized (i.e. stripping ../ ./) absolute path from static directory
    const localFile = path.normalize(`${staticDir}/${file}`)
    // We only serve files from staticDir folder no path traversal
    if(localFile.startsWith(staticDir) && localFile !== staticDir) {
      fs.access(localFile, fs.constants.R_OK, (err) => {
        if (!err) {
          resolve(file);
        } 
        resolve(null);
      });
    } else {
      resolve(null);
    }
  });
}

/**
 * Verify we have the requested file in stock
 * If we have, but with the wrong path, do a redirect
 * If not return the modified index.html
 * @param {Express.Request} req
 * @param {Express.Response} res
 * @param {(err?: any) => void} next
 */
function redirectMiddleware(req, res, next) {
  const reqURL = req.url;
  foundFileFromPath(reqURL).then((url) => {
      if(url) {
        // with url rewrite
        req.url = `/${url}`
        return next();
        // // with redirection
        // if (reqURL === `/${url}`) {
        //   // Serve the file with 'serve-static'
        //   return next();
        // }
        // return res.redirect(`/${url}`);        
      }
      // Not found, send the modified index.html
      res.setHeader('Cache-Control', 'public, max-age=0');
      res.send(res.locals.htmlWithNonces);
  });
}
```

Reply:
The idea was to be able to deploy blip under a sub directory.

Reply:
For now we don't, and if we had to we would have just one level down to search (and we would modify the staticDir variable)